### PR TITLE
Endline Detection

### PR DESCRIPTION
**Pull request by w29ahmed**
I'll add some comments here later

- Allan is gonna add a README to describe the new (and improved) algorithm for this.
- Tested with a ros bag and it works really well!

### CODE REVIEW

File: rr_computer_vision/include/endline_detection.hpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#ifndef ENDLINE_DETECTION_HPP
+#define ENDLINE_DETECTION_HPP
+
+#include <ros/ros.h>
+#include <image_transport/image_transport.h>
+#include <cv_bridge/cv_bridge.h>
+#include <sensor_msgs/image_encodings.h>
+#include <opencv2/opencv.hpp>
+#include <std_srvs/Trigger.h>
+// #include "msg_srv_names.hpp"
```
Comment (DongJunJin) [inline rr_computer_vision/include/endline_detection.hpp]: Please remove this if this is not used anywhere

File: rr_computer_vision/include/endline_detection.hpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  public :
+    EndlineDetection(ros::NodeHandle);
+    void ImgCb(const sensor_msgs::ImageConstPtr&);
+    static bool compareContourAreas( std::vector<cv::Point> contour1, std::vector<cv::Point> contour2 );
+    
+  private :
+    ros::NodeHandle nh_;
+    ros::ServiceClient client_;
+    image_transport::ImageTransport it_;
+    image_transport::Subscriber test_subscriber;
```
Comment (DongJunJin) [inline rr_computer_vision/include/endline_detection.hpp]: change the variable names to something with context; this doesnt tell me the purpose of what these are

File: rr_computer_vision/include/endline_detection.hpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    image_transport::Publisher test_publisher;
+
+    bool detection_status;
```
Comment (DongJunJin) [inline rr_computer_vision/include/endline_detection.hpp]: Inconsistent naming convention; please change these to be consistent with the others

File: rr_computer_vision/launch/computer_vision_base.launch [hunk 4]
```diff
@@ -7,9 +7,6 @@
 <launch>
   <arg name="camera" default=""/>
   <arg name="config_file" default=""/>
-    
-  <!-- Launch ZED camera wrapper -->
-  <include file="$(find zed_wrapper)/launch/zed.launch"> </include>
```
Comment (DongJunJin) [inline rr_computer_vision/launch/computer_vision_base.launch]: Why did you remove the zed? does the computer vision base not need zed?

File: rr_computer_vision/src/computer_vision_main.cpp [hunk 5]
```diff
@@ -19,7 +18,9 @@ int main(int argc, char** argv)
   ros::init(argc, argv, "rr_computer_vision");
   ros::NodeHandle nh;
   ROS_INFO("Starting Computer Vision Node");
+  
   ComputerVision computer_vision(nh);
+  // EndlineDetection ec(nh);
```
Comment (DongJunJin) [inline rr_computer_vision/src/computer_vision_main.cpp:23]: Why is this commented out if endline works?

File: rr_computer_vision/src/endline_detection.cpp [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+#include "endline_detection.hpp"
+#include "opencv2/highgui/highgui.hpp"
+#include "opencv2/imgproc/imgproc.hpp"
+#include <vector>
+
+// Constructor
+EndlineDetection::EndlineDetection(ros::NodeHandle nh) : it_(nh_)
+{
+  detection_status = false;
```
Comment (DongJunJin) [inline rr_computer_vision/src/endline_detection.cpp]: declare these boys on how you declared the nodehandle; this will provide for a more consistent look on the code

File: rr_computer_vision/src/endline_detection.cpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  endline_counter = 0;
+  client_ = nh_.serviceClient<std_srvs::Trigger>("/Supervisor/count_lap");
+  
+  test_subscriber = it_.subscribe("/zed/rgb/image_rect_color", 1, &EndlineDetection::ImgCb, this);
+  // test_publisher = it_.advertise("/test_endline", 1);
```
Comment (DongJunJin) [inline rr_computer_vision/src/endline_detection.cpp]: please remove unused lines

File: rr_computer_vision/src/endline_detection.cpp [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+#include <vector>
+
+// Constructor
+EndlineDetection::EndlineDetection(ros::NodeHandle nh) : it_(nh_)
+{
+  detection_status = false;
+  endline_counter = 0;
+  client_ = nh_.serviceClient<std_srvs::Trigger>("/Supervisor/count_lap");
+  
+  test_subscriber = it_.subscribe("/zed/rgb/image_rect_color", 1, &EndlineDetection::ImgCb, this);
```
Comment (DongJunJin) [inline rr_computer_vision/src/endline_detection.cpp]: Function name is image callback but what is this callback processing? Make the name clearer so someone doesnt have to read the inside to figure out what it does

File: rr_computer_vision/include/endline_detection.hpp [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+#include <opencv2/opencv.hpp>
+#include <std_srvs/Trigger.h>
+// #include "msg_srv_names.hpp"
+
+class EndlineDetection
+{
+  public :
+    EndlineDetection(ros::NodeHandle);
+    void ImgCb(const sensor_msgs::ImageConstPtr&);
+    static bool compareContourAreas( std::vector<cv::Point> contour1, std::vector<cv::Point> contour2 );
```
Comment (DongJunJin) [inline rr_computer_vision/include/endline_detection.hpp]: Genuine question because i dont know. Why is this a static function?

File: rr_computer_vision/src/computer_vision_main.cpp (hunk 5 above)
Comment (w29ahmed) [reply rr_computer_vision/src/computer_vision_main.cpp:23]: Cuz in this main function, everything should eventually be setup through a rosparam indicating the race type, and that will instantiate the necessary objects for that race. Also to avoid overhead of all the callbacks from this running when people branch out from develop

File: rr_computer_vision/src/endline_detection.cpp (hunk 7 above)
Comment (w29ahmed) [reply rr_computer_vision/src/endline_detection.cpp]: I think it's useful to leave it here commented out. Would be useful for somebody in the future who wants to quickly see the result, instead of having to basically rewrite this code

File: rr_computer_vision/include/endline_detection.hpp (hunk 9 above)
Comment (w29ahmed) [reply rr_computer_vision/include/endline_detection.hpp]: It's used as a lambda function to sort the contour areas, and those need to be static otherwise compiler complains

File: rr_computer_vision/include/endline_detection.hpp (hunk 1 above)
Comment (w29ahmed) [reply rr_computer_vision/include/endline_detection.hpp]: Had it here cuz wanted to put the supervisor service callbacks into that common namespace and reference them in here that way instead of hardcoded strings. But this is extra work and Ima just put it off tbh cuz who really cares

**Review by DongJunJin — APPROVED** [review]
